[terra-show-hide] prop to focus on the element added (#3838)
cerner/terra-core: packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx

import { Badge } from 'terra-show-hide/package.json?dev-site-package';
import { Notice } from "@cerner/terra-docs";

import DefaultShowHide from './example/DefaultShowHide?dev-site-example';
import CustomButtonTextShowHide from './example/CustomButtonTextShowHide?dev-site-example';
import LinkShowHide from './example/LinkShowHide?dev-site-example';
import InitiallyOpenShowHide from './example/InitiallyOpenShowHide?dev-site-example';
import ButtonAlignCenterShowHide from './example/ButtonAlignCenterShowHide?dev-site-example';
import ButtonAlignRightShowHide from './example/ButtonAlignRightShowHide?dev-site-example';
import NoPreviewShowHide from './example/NoPreviewShowHide?dev-site-example';
import NoFocusRefShowHide from './example/NoFocusRefShowHide?dev-site-example';

import ShowHidePropsTable from 'terra-show-hide/src/ShowHide?dev-site-props-table';
import ShowHideFocuserPropsTable from 'terra-show-hide/src/ShowHideFocuser?dev-site-props-table';

<Badge />

# Terra Show Hide

> Show Hide Component - is a disclosure widget that enables content to be either collapsed (hidden) or expanded (visible). It has two elements: a disclosure [button][1] and a section of content whose visibility is controlled by the button. When the controlled content is hidden, the button is often styled as a typical push button with a down-pointing caret to hint that activating the button will display additional content. When the content is visible, the content displays above the button and the caret points up.

Author: fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

> Another Terra component, [Toggle Button][2], is available and preferred to Show/Hide because it better adheres to accessibility best practices for disclosure components. It differs from Show Hide by displaying the content below the button when it is activated. 

Reviewer: ```suggestion
Another Terra component, [Toggle Button][2], is available and preferred to Show/Hide when possible because it better adheres to accessibility best practices for disclosure components. It differs from Show Hide by displaying the content below the button when it is activated. 
```

## Usage

```jsx
import ShowHide, { Paragraph } from 'terra-show-hide';
```

#### A single paragraph of text
  - At the top level of your component declare a `ref`, pass it as `focusRef` props to `ShowHide` component. Pass a preview node as a prop to `ShowHide` component.
  - Use `Paragraph` component inside `<p>` tag. A `Paragraph` component allows `ShowHide` component to set focus on the beginning of the hidden portion of the text upon opening. 
  - The text splitted in the preview and the rest of the text portions has to be passed as string props, and the `focusRef` should be forwarded to `Paragraph` component.

  ```jsx
  import React, { useRef, useState } from 'react';
  import ShowHide, { Paragraph } from 'terra-show-hide';

  const MyComponent = () => {
    const focusRef = useRef(null);
    const [isOpen, setIsOpen] = useState(false);
    const toggleShowHide = () => setIsOpen(!isOpen);

>     // Split text in the preview and the rest of the text portions.

Author: @eawww I can reproduce that. It only happens in <li> example, the JAWS gets confused with the previous example if it is open, but for list example only. Please test the list example with the previous example close for now, I am working on figuring out what causes this bug and how to avoid it.

>     // Split text in the preview and the rest of the text portions.

Author: Final fix for JAWS and VO confusion between multiple examples: https://github.com/cerner/terra-core/pull/3838/commits/ef2f17da26d15791ecac86c8dde5f0b911c768d4

    const previewText = 'Patients are requesting greater affordability and efficiency in healthcare.';
    const text = 'With procedures performed in an ambulatory surgery center costing up to 60% less compared to a hospital outpatient department (1), the demand for these facilities is increasing.';

    return (
      <ShowHide focusRef={ focusRef } preview={ <p>{ previewText }</p> } isOpen={isOpen} onChange={toggleShowHide} >
        <p>
          <Paragraph ref={ focusRef } previewText={ previewText } text={ text } />
        </p>
      </ShowHide>
    );
  }
  ```

#### Multiple children
  - At the top level of your component declare a `ref`, pass it as `focusRef` props to `ShowHide` component. Pass a preview node as a prop to `ShowHide` component.
  - A `Paragraph` component allows `ShowHide` component to set focus on the beginning of the hidden portion of the text upon opening. 

>   - Dublicate the preview nodes as `ShowHide` children.

Author: Fixed in https://github.com/cerner/terra-core/commit/0156351be0b642aea10b4a80d2c9297b2d074298

      const [isOpen, setIsOpen] = useState(false);
      const toggleShowHide = () => setIsOpen(!isOpen);

      // This example's content is a header and a list of 4 items.
      const header = 'Key Benefits of Cerner Cardiovascular Solutions';
      const list = [
        'Eliminates silos of information and the resulting inefficiencies with a unified EHR and cardiovascular system',
        'Enhances clinical, financial and performance outcomes with comprehensive procedural documentation',
        'Creates efficient diagnostic workflows, image management and analysis',
        'Promotes cardiac disease management through health maintenance protocols',
      ];

      // The header and the first two items of the list will be displayed in preview portion.
      const preview = [
        <h3 key="header">{header}</h3>,
        <ul key="list">
          <li>{list[0]}</li>
          <li>{list[1]}</li>
        </ul>,
      ];

      return (
        <ShowHide focusRef={focusRef} preview={preview} isOpen={isOpen} onChange={toggleShowHide} >
          <h3>{header}</h3>
          <ul>
            <li>{list[0]}</li>
            <li>{list[1]}</li>
            <li>

>               <Paragraph ref={focusRef} text={list[2]} />

Author: No, list[3] won't be visible. The showHide component was initially designed in such a way, that only what passed as a preview prop will be visible. In this case the preview portion is header, list[1] and list[2]:
https://github.com/cerner/terra-core/blob/74096b11bbe9ed5372617a339128bc531b7bac2b/packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx#L106-L112

            </li>
            <li>{list[3]}</li>
          </ul>
        </ShowHide>
      );
    }
  ```

#### First hidden child is an interactive element (link, button, etc)
  - At the top level of your component declare a `ref`.
  - Forward `focusRef` to the interactive element directly. There is no need to use `Paragraph` in this case.

  ```jsx
    import React, { useState, useRef } from 'react';
    import ShowHide from 'terra-show-hide';

    const MyComponent = () => {
      const focusRef = useRef(null);
      const [isOpen, setIsOpen] = useState(false);
      const toggleShowHide = () => setIsOpen(!isOpen);

      const previewText = 'Patients are requesting greater affordability and efficiency in healthcare.';

      return (
        <ShowHide focusRef={ focusRef } preview={ <p>{ previewText }</p> } isOpen={isOpen} onChange={toggleShowHide} >
          <p>{ previewText }</p>
          <a ref={this.focusRef} href="https://www.cerner.com/" alt="Oracle Cerner">Oracle Cerner</a>
        </ShowHide>
      );
    }
  ```

## Accessibility

<Notice variant="important" ariaLevel="3">

> In order for the component to be accessible, it's important to follow the inplimentation guidance. Even though the focusRef prop is not required for backward compatibility, a failure to use it will result in lack of accessibility.

Author: Fixed in https://github.com/cerner/terra-core/pull/3838/commits/0156351be0b642aea10b4a80d2c9297b2d074298

## Usage

```jsx
import ShowHide, { Focuser } from 'terra-show-hide';
```

> ## Accessibility

Reviewer 2: Wonder if this section would be better placed after the "Examples" section that explains the basic usage.

Author: Thank you, I am also not sure where to place that block. We might want to re-visit that question after the rest of documentation is approved.

<Notice variant="important" ariaLevel="3">

  To make the Show/Hide component accessible, you must use the `focusRef` prop. This prop allows assistive technologies such as screen readers to put focus on the hidden text and announce it when it is disclosed. The `focusRef` is optional for backward compatibility but is required to comply with accessibility requirements.

</Notice>

> #### Truncate a paragraph of text

Reviewer: Even though the Notice above gives an h3, this feels like these headings should be level 3 as well since the header in the notice should only apply to what's grouped within it.

Author: Thank you! Changed the headings level to be level 3: https://github.com/cerner/terra-core/pull/3838/commits/09d4e4543413b7aa554d30b921cfe7f1ec1455b4

>   - At the top level of your component declare a `ref`, pass it as `focusRef` props to `ShowHide` component. 

Reviewer: ```suggestion
  - At the top level of your component declare a `ref`, pass it as `focusRef` prop to `ShowHide` component. 
```

>   - Pass a preview node with the portion of the text as a prop to `ShowHide` component.

Reviewer: ```suggestion
  - Pass a node with the preview portion of the text as a prop to `ShowHide` component.
```

>   - Use `Focuser` component inside `<p>` tag. A `Focuser` component allows `ShowHide` component to set focus on the beginning of the focusable text upon opening. 

Reviewer: ```suggestion
  - Use the `Focuser` component inside `<p>` tag. A `Focuser` component allows `ShowHide` component to set focus on the beginning of the focusable text upon opening. 
```

>   - The text splitted in the prefix and focusable text portions has to be passed as string props. The prefix prop should contain the portion of the text that was a part of the preview prop, passed to `ShowHide`.

Author: Good catch, appreciate it. Fixed in https://github.com/cerner/terra-core/pull/3838/commits/09d4e4543413b7aa554d30b921cfe7f1ec1455b4

    const focusRef = useRef(null);
    const [isOpen, setIsOpen] = useState(false);
    const toggleShowHide = () => setIsOpen(!isOpen);

    const prefix = 'Patients are requesting greater affordability and efficiency in healthcare. With procedures performed in an ambulatory surgery center costing up to 60% less compared to a hospital outpatient department (1), the demand for these facilities is increasing.';
    const focusableText = 'In fact, the U.S. ambulatory surgery center market is expected to see a 6.9 % compound annual growth rate, reaching $33 billion by 2028. (2) Cerner understands the urgency to grow in the ambulatory surgery center market while continuing to deliver excellent care. Healthcare IT products can help improve clinician efficiency and patient outcomes, as well as enhance communication and data exchange between ambulatory surgery center providers and patients. Read';

    return (
      <ShowHide focusRef={focusRef} preview={ <p>{prefix}</p> } isOpen={isOpen} onChange={toggleShowHide} >
        <p>
          <Focuser ref={this.focusRef} prefix={prefix} focusableText={focusableText} />
        </p>
      </ShowHide>
    );
  }
  ```

#### Multiple children

>   Even though the main purpose of the `ShowHide` component is to truncate long paragraps of texts, the content of the `ShowHide` does not have to be one paragraph. It can include links, lists, and any king of HTML tags as long as the first line of the hidden text is passed as a string prop to `Focuser` component rather than an HTML element. Please note:
>   - Links can be placed inside the `p` tag that wrapps `Focuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passes as an HTML element preceding `Focuser`.

Reviewer: ```suggestion
  Even though the main purpose of the `ShowHide` component is to truncate long paragraps of texts, the content of the `ShowHide` does not have to be one paragraph. It can include links, lists, and any kind of HTML tags as long as the first line of the hidden text is passed as a string prop to `Focuser` component rather than an HTML element.
  
Please note:
  - Links can be placed inside the `p` tag that wraps `Focuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passed as an HTML element preceding `Focuser`.
```

>   - Everythin that precedes `Focuser` component must be also included into the `ShowHide` preview prop.

Reviewer 2: sp: "Everything"

Author: Good catch, appreciate it. Fixed in https://github.com/cerner/terra-core/pull/3838/commits/09d4e4543413b7aa554d30b921cfe7f1ec1455b4

>   - Do not break lists placing `Focuser` inside `li`, as it might break the assistive technologies reading order.

Reviewer: ```suggestion
  - Do not break lists placing `Focuser` inside `li`, as it might cause assistive technologies to read the content incorrectly.
```

  - Enter key activates the disclosure control and toggles the visibility of the disclosure content.
  - Space key activates the disclosure control and toggles the visibility of the disclosure content. 

### Truncate a paragraph of text

  - At the top level of your component declare a `ref`, pass it as `focusRef` prop to `ShowHide` component.
  - Pass a node with the preview portion of the text as a prop to `ShowHide` component.
  - Use the `ShowHideFocuser` component inside `<p>` tag. A `ShowHideFocuser` component allows `ShowHide` component to set focus on the beginning of the focusable text upon opening.
  - The text split in the prefix and focusable text portions must be passed as string props. The prefix prop should contain the portion of the text that was a part of the preview prop, passed to `ShowHide`.
  - Forward `focusRef` to `ShowHideFocuser` component.

  ```jsx
  import React, { useRef, useState } from 'react';
  import ShowHide, { ShowHideFocuser } from 'terra-show-hide';

  const MyComponent = () => {
    const focusRef = useRef(null);
    const [isOpen, setIsOpen] = useState(false);
    const toggleShowHide = () => setIsOpen(!isOpen);

>     const prefix = 'Patients are requesting greater affordability and efficiency in healthcare. With procedures performed in an ambulatory surgery center costing up to 60% less compared to a hospital outpatient department (1), the demand for these facilities is increasing.';

Reviewer: ```suggestion
    const prefix = 'Patients are requesting greater affordability and efficiency in healthcare. With procedures performed in an ambulatory surgery center costing up to 60% less compared to a hospital outpatient department (1), the demand for these facilities is increasing...';
```

  - Everything that precedes `ShowHideFocuser` component must be also included into the `ShowHide` preview prop.
  - Do not break lists placing `ShowHideFocuser` inside `li`, as it might cause assistive technologies to read the content incorrectly.

  ```jsx
    import React, { useState, useRef } from 'react';
    import ShowHide, { ShowHideFocuser } from 'terra-show-hide';

    const MyComponent = () => {
      const focusRef = useRef(null);
      const [isOpen, setIsOpen] = useState(false);
      const toggleShowHide = () => setIsOpen(!isOpen);

      const header = 'Bringing clarity to life sciences and healthcare decision-making ';
      const prefix = 'When we accelerate the discovery, development and deployment of life-enhancing insights and therapies, clinicians and patients benefit.';
      const preview = [
        <h3>{header}</h3>,
        <p>{prefix}</p>,
      ];

      return (
        <ShowHide focusRef={focusRef} preview={preview} isOpen={isOpen} onChange={toggleShowHide} >
          <h3>{header}</h3>
          <p>

>             <ShowHideFocuser ref={this.focusRef} prefix={prefix} focusableText="Several factors help bring clarity to life sciences, researchers and clinicians and " />

Reviewer: this could use a comment describing exactly what the resulting behavior of this would be.

Author: Adde in https://github.com/cerner/terra-core/pull/3838/commits/1813b072e8074cbe8d2ec0d737f42a88b9b81774

## Usage

```jsx
import ShowHide, { ShowHideFocuser } from 'terra-show-hide';
```

## Accessibility

<Notice variant="important" ariaLevel="3">

>   To make the Show/Hide component accessible, you must use the `focusRef` prop. This prop allows assistive technologies such as screen readers to move user focus backward to hidden text and announce it when it is disclosed. For backward compatibility, the `focusRef` prop is optional, but is required to meet accessibility requirements.

Author: fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

</Notice>

<Notice variant="caution" ariaLevel="3">

>   This component is not recommended for use in Safari due to VoiceOver focus limitations with Show Hide. Consider using Terra Toggle Button instead.

Author: fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

</Notice>

###  Keyboard Interactions 
  When the disclosure control has focus:
  - Enter key activates the disclosure control and toggles the visibility of the disclosure content.

>   - Space key activates the disclosure control and toggles the visibility of the disclosure content. 

Author: I guess @mjpalazzo put it that way in purpose. He can answer that better than me.

> ### Truncate a paragraph of text

Author: Fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

>   - At the top level of your component declare a `ref`, pass it as `focusRef` prop to `ShowHide` component.

Author: Fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

  - Pass a node with the preview portion of the text as a prop to `ShowHide` component.
  - Use the `ShowHideFocuser` component inside `<p>` tag. A `ShowHideFocuser` component allows `ShowHide` component to set focus on the beginning of the focusable text upon opening.
  - The text split in the prefix and focusable text portions must be passed as string props. The prefix prop should contain the portion of the text that was a part of the preview prop, passed to `ShowHide`.
  - Forward `focusRef` to `ShowHideFocuser` component.

  ```jsx
  import React, { useRef, useState } from 'react';
  import ShowHide, { ShowHideFocuser } from 'terra-show-hide';

  const MyComponent = () => {
    const focusRef = useRef(null);
    const [isOpen, setIsOpen] = useState(false);
    const toggleShowHide = () => setIsOpen(!isOpen);

    const prefix = 'Patients are requesting greater affordability and efficiency in healthcare. With procedures performed in an ambulatory surgery center costing up to 60% less compared to a hospital outpatient department (1), the demand for these facilities is increasing.';

>     const focusableText = 'In fact, the U.S. ambulatory surgery center market is expected to see a 6.9 % compound annual growth rate, reaching $33 billion by 2028. (2) Cerner understands the urgency to grow in the ambulatory surgery center market while continuing to deliver excellent care. Healthcare IT products can help improve clinician efficiency and patient outcomes, as well as enhance communication and data exchange between ambulatory surgery center providers and patients. Read';

Author: removed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

    return (
      <ShowHide focusRef={focusRef} preview={ <p>{prefix}</p> } isOpen={isOpen} onChange={toggleShowHide} >
        <p>
          <ShowHideFocuser ref={this.focusRef} prefix={prefix} focusableText={focusableText} />
        </p>
      </ShowHide>
    );
  }
  ```

### Multiple children

>   Even though the main purpose of the `ShowHide` component is to truncate long paragraps of texts, the content of the `ShowHide` does not have to be one paragraph. It can include links, lists, and any kind of HTML tags as long as the first line of the hidden text is passed as a string prop to `ShowHideFocuser` component rather than an HTML element.

Author: fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

  
Please note:

>   - Links can be placed inside the `p` tag that wraps `ShowHideFocuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passed as an HTML element preceding `ShowHideFocuser`.

Author: fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

>   - Everything that precedes `ShowHideFocuser` component must be also included into the `ShowHide` preview prop.

Author: fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

>   - Do not break lists placing `ShowHideFocuser` inside `li`, as it might cause assistive technologies to read the content incorrectly.

Author: fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

      const preview = [
        <h3>{header}</h3>,
        <p>{prefix}</p>,
      ];

      return (
        <ShowHide focusRef={focusRef} preview={preview} isOpen={isOpen} onChange={toggleShowHide} >
          <h3>{header}</h3>
          // This paragraph contains a link in hidden content. Explore a working example below under Examples, Custom Button Text ShowHide.
          <p>
            <ShowHideFocuser ref={this.focusRef} prefix={prefix} focusableText="Several factors help bring clarity to life sciences, researchers and clinicians and " />
            <a href="#">guide better care</a>, including:
          </p>
          <ul>
            <li>Data-driven research solutions for commercial and real-world evidence collaborators and health providers</li>
            <li>Support to move from simply accumulating and analyzing data to linking and generating deeper insights </li>
            <li>Combined life sciences knowledge, healthcare innovation and collaborative research expertise</li>
          </ul>
        </ShowHide>
      );
    }
  ```

### First hidden child is an interactive element (link, button, etc)

>   - At the top level of your component declare a `ref`.

Author: fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1
